Fst2List.cpp

```diff
@@ -1243,6 +1347,330 @@ class CFstApp {
+    }
+    a->initial_states[a->number_of_graphs] = a->number_of_states;
+    a->number_of_states_per_graphs[a->number_of_graphs] = 2;
+    a->number_of_states += 2;
+    a->states[a->number_of_states - 2] = new_Fst2State(NULL);
+    set_initial_state(a->states[a->number_of_states - 2], 1);
+    a->states[a->number_of_states - 1] = new_Fst2State(NULL);
+    set_final_state(a->states[a->number_of_states -1], 1);
+    int last_number_of_tags = a->number_of_tags;
+    a->number_of_tags += processedLexicalMasks[lexicalMaskCnt].entriesCnt;
+    a->tags = (Fst2Tag*)realloc(a->tags, a->number_of_tags * sizeof(Fst2Tag));
+    if(a->tags == NULL) {
+      fatal_error("realloc error for tags in createLexicalMaskSubgraph");
+    }
+    // create a new tag for each entry found in processedLexicalMask at lexicalMaskCnt index
+    int k = a->number_of_tags - 1;
+    for(int i = last_number_of_tags; i < a->number_of_tags; i++) {
+      a->tags[i] = new_Fst2Tag(NULL);
+      a->tags[i]->input = u_strdup(processedLexicalMasks[lexicalMaskCnt].entries[k - last_number_of_tags].input);
+      if(processedLexicalMasks[lexicalMaskCnt].lexicalMask.output != NULL) {
+        a->tags[i]->output = u_strdup(processedLexicalMasks[lexicalMaskCnt].lexicalMask.output);
+      }
+      add_transition_to_state(a->states[a->number_of_states - 2], i, a->number_of_states - 1, NULL);
+      k--;
+    }
+  }
+
+  /**
+  * count the number of lexical mask in the current automaton
+  */
+  int countLexicalMasks() {
+    int count = 0;
+    for(int j = 0; j < a->number_of_states; j++) {
+      Transition *t = a->states[j]->transitions;
+      while(t != NULL) {
+        if(!(t->tag_number & SUBGRAPH_PATH_MARK) && (a->tags[t->tag_number]->input[0] == '<' && a->tags[t->tag_number]->input[u_strlen(a->tags[t->tag_number]->input) - 1] == '>')
+          && u_strcmp(a->tags[t->tag_number]->input, "<E>")) {
+          count++;
+        }
+        t = t->next;
+      }
+    }
+    return count;
+  }
+  
+  /**
+   * lexical masks which should be read literally and not be regarded as lexical masks which should be extracted into binary dictionaries
+   */
+  bool maskMustBeReadLiterally(unichar *lexical_mask){
+	 char special_mask[NB_LITTERAL_MASKS][16] = {"<NB>", "<PRE>", "<MIX>", "<$>", "<^>", "<MAJ>", "<PNC>", "<MOT>",
+		                           "<MIN>", "<WORD>", "<FIRST>", "<LOWER>", "<UPPER>", "<TOKEN>", "<LETTER>"}; // 4.3 manual
+     if (lexical_mask == NULL || u_starts_with(lexical_mask, "<!")){ // negative lexicals masks are processed literally too
+		 return true;
+	 }
+	 for (int i = 0; i < NB_LITTERAL_MASKS; i++){
+         if (!u_strcmp(lexical_mask, special_mask[i])){
+            return true;
+         }
+     }
+     return false;
+  }
+  
+  /**
+   * tells if the lexical masks is a DIC mask
+   * DIC masks require the extraction of all entries from binary dictionaries
+   */
+  bool isDicMask(unichar *lexical_mask){
```

Comment:
Add comment:
This is a functional difference with LocatePattern
LocatePattern matches <SDIC> only with simple entries and <CDIC> only with multiword entries